### ISSUE TITLE
404 Not Found - Browser Gets Forwarded to /404.xhtml

### ISSUE THREAD

**Issue by mheppler** [IQSS/dataverse#2367]
When looking at how our site handles URLs that are not found -- AKA, 404 Not Found Error -- I noticed that the bad URL is not retained in the browser window, instead it gets forwarded to the URL /404.xhtml. Other sites don't forward you, so that you can see what the bad URL is -- maybe it was a typo, and you can easily fix it.

![screen shot 2015-07-16 at 4 58 26 pm](https://cloud.githubusercontent.com/assets/687227/8735049/eda24f14-2bdb-11e5-8d2f-4bf5518accdc.png)

Related(?) to #1057.

**Comment by eaquigley** [comment]
+1

**Comment by mheppler** [comment]
Turned to StackOverflow in a quest for answers.

[PrettyFaces 404 page without URL redirect http://stackoverflow.com/questions/31574822/prettyfaces-404-page-without-url-redirect]

**Comment by pdurbin** [comment]
The [answer](http://stackoverflow.com/questions/31574822/prettyfaces-404-page-without-url-redirect/31589174#31589174) by @lincolnthree who created PrettyFaces is obviously worth checking out!

**Comment by bencomp** [comment]
This is not just about seeing the wrong URL - it breaks semantics of the web, because instead of giving an error response to the browser, this is just showing the error text. The browser is told to find the answer to a bogus request at /404.xhtml and because that is there, the browser 'thinks' the request was handled correctly (`200 OK`). If a browser were to cache content, it might cache the error page as a valid response.

This goes for 403 responses too. Please put this issue on the roadmap.

**Comment by lincolnthree** [comment]
Yes, regarding the HTTP status code, you should still send '404 NOT FOUND' when forwarding to the error page.

**Comment by mheppler** [comment]
This is also happening for the 403 pg, as found in #2838.

**Comment by scolapasta** [comment]
This has been fixed in the widgets branch.

**Comment by pdurbin** [comment]
@sekmiller showed me how this works with "/loginpage.xhtml" for example and how the Shib code should be modified after edf12fb and related commits are merged into "develop". I asked him not to touch the Shib code in the widgets branch since it will only lead to a merge conflict on pull request #3025. The bottom line for @kcondon is that this might need to be retested as we integrate multiple branches into "develop".

**Comment by pdurbin** [comment]
@scolapasta if https://github.com/IQSS/dataverse/tree/2469-widgets gets merged before https://github.com/IQSS/dataverse/tree/2939-shib (pull request #3025) shouldn't a new issue get opened to have the Shib code use the new sendError method that will only be available once the 2469-widgets branch is merged? I'm asking because at https://github.com/IQSS/dataverse/pull/3025#issuecomment-218509385 @eaquigley wrote "Good on my end. Passing to @kcondon to merge."

**Comment by kcondon** [comment]
Works, closing, error is presented but URL is preserved and correct http code is returned.
